File: src/main/scala/top/criwits/scaleda/verilog/psi/nodes/module/PortDeclarationPsiNode.java
package top.criwits.scaleda.verilog.psi.nodes.module;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import org.antlr.intellij.adaptor.psi.ANTLRPsiNode;
import org.jetbrains.annotations.NotNull;
import top.criwits.scaleda.verilog.psi.nodes.TypedDeclaration;

import java.util.Arrays;
import java.util.stream.Collectors;

public class PortDeclarationPsiNode extends ANTLRPsiNode
        implements TypedDeclaration {

    public PortDeclarationPsiNode(@NotNull ASTNode node) {
        super(node);
    }

    /**
     * in {input|output|inout}_declaration children except last child is about type,
     * e.g. "output wire signed [15:0] OUTPUT_NAME", so we will pass it to typeText
     */
    @Override
    public String getTypeText() {
        PsiElement[] typeChildren = Arrays.copyOf(
                this.getLastChild().getChildren(),
                this.getLastChild().getChildren().length - 1
        );
        return String.join(" ",
                Arrays.stream(typeChildren)
                        .map(PsiElement::getText)
                        .collect(Collectors.toList())
        );
    }

}

File: src/main/scala/top/criwits/scaleda/verilog/psi/nodes/signal/PortDeclarationPsiNode.scala
package top.criwits.scaleda
package verilog.psi.nodes.signal

import verilog.psi.nodes.signal.PortDeclarationPsiNode.{INOUT, INPUT, OUTPUT, OUTPUT_REG}

import com.intellij.lang.ASTNode
import com.intellij.psi.util.PsiTreeUtil

class PortDeclarationPsiNode(node: ASTNode) extends SignalDeclarationPsiNode(node) {
  def getDeclaration: Option[AbstractDeclarationPsiNode] = {
    Option(PsiTreeUtil.getChildOfType(this, classOf[AbstractDeclarationPsiNode]))
  }

  def getPortType: PortDeclarationPsiNode.PortType = {
    val typeText = getTypeText
    if (typeText.contains("input")) INPUT
    else if (typeText.contains("inout")) INOUT
    else if (typeText.contains("output")) OUTPUT
    else if (typeText.contains("output") && (typeText.contains("reg") || typeText.contains("integer") || typeText.contains("time"))) OUTPUT_REG
    else INPUT // default
  }

  override def getTypeText: String = {
    val declaration = PsiTreeUtil.getChildOfType(this, classOf[AbstractDeclarationPsiNode])
    if (declaration == null) "unknown" else declaration.getTypeText
  }

  override def getRange: Option[RangePsiNode] = getDeclaration.flatMap(d => Option(PsiTreeUtil.getChildOfType(d, classOf[RangePsiNode])))
}

object PortDeclarationPsiNode extends Enumeration {
  type PortType = Value
  val INPUT, INOUT, OUTPUT, OUTPUT_REG = Value
}
